libraries/GCS_MAVLink/GCS.h

```diff
@@ -1288,6 +1293,7 @@ class GCS
 private:
 
     static GCS *_singleton;
+    HAL_Semaphore _sem;
```

Comment:
I'm not sure if this is insufficent protection or too blunt.

It's too blunt if you're protecting against mavlink-command-ints being run at the same time - that should go within `handle_command_int`.

It might not be sufficient protection depending on where the main thread is up to when the script command is executed.  Remembering, of course, that we process mavlink while we're sleeping (at least that's only likely when we're disarmed....)

Reply:
I presume you mean in AP_Vehicle::scheduler_delay_callback() ? that calls gcs().update_receive() which is exactly the call the mutex in this PR protects.

Reply:
Upon further thought I agree that this is insufficient locking. Consider, for example `AP_Gripper_Servo::release`. It can be called through `GCS_MAVLINK::handle_command_do_gripper` which is called by `update_receive` and is protected by that new lock. But on the same object `AP_Gripper_Servo::grab` can be called through `AP_Gripper_Backend::update()` which is called by the main scheduler on the main thread too and not protected by any lock.

The scheduler won't run two things at once (IIUC) but if Lua is releasing while the main thread decides to grab, who wins? We need another lock in the gripper methods too. Then we have to start being concerned about locking order and deadlocks, and look at locking in everything that can be done by the command int handler...

This problem for Gripper is already triggerable through the aux function handler and bindings in any case (though adding a generic lock there would be a lot more straightforward), but I don't think we want it to be worse.